pkg/validation/cluster.go

```diff
@@ -0,0 +1,129 @@
+package validation
+
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/kubelet/config/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/k8stopologyawareschedwg/deployer/pkg/validator"
+
+	mcov1 "github.com/openshift/machine-config-operator/pkg/apis/machineconfiguration.openshift.io/v1"
+
+	"github.com/openshift-kni/numaresources-operator/pkg/kubeletconfig"
+)
+
+func ValidateCluster(aContext context.Context, aClient client.Client) ([]validator.ValidationResult, error) {
+
+	var result []validator.ValidationResult
+
+	mcoKubeletConfigList := mcov1.KubeletConfigList{}
+	if err := aClient.List(aContext, &mcoKubeletConfigList); err != nil {
+		return nil, err
+	}
+
+	for _, mcoKubeletConfig := range mcoKubeletConfigList.Items {
+		kubeletConfig, err := kubeletconfig.MCOKubeletConfToKubeletConf(&mcoKubeletConfig)
+		if err != nil {
+			return nil, err
+		}
+
+		var nodeNames []string
+		machineConfigPools, err := getMCPsFromMCOKubeletConfig(aContext, aClient, mcoKubeletConfig)
+		if err != nil {
+			return nil, err
+		}
+		for _, mcp := range machineConfigPools {
+
+			nodes, err := getNodeListFromMachineConfigPool(aContext, aClient, *mcp)
+			if err != nil {
+				return nil, err
+			}
+			nodeNames = getNodeNames(nodes)
+		}
+
+		for _, nodename := range nodeNames {
+			ret := validateKubeletConfig(nodename, *kubeletConfig)
+			result = append(result, ret...)
+		}
+
+	}
+	return result, nil
+}
+
+func getNodeListFromMachineConfigPool(aContext context.Context, aClient client.Client, mcp mcov1.MachineConfigPool) ([]corev1.Node, error) {
+	sel, err := metav1.LabelSelectorAsSelector(mcp.Spec.MachineConfigSelector)
+	if err != nil {
+		return nil, err
+	}
+
+	nodeList := &corev1.NodeList{}
+	err = aClient.List(aContext, nodeList, &client.ListOptions{LabelSelector: sel})
+	if err != nil {
+		return nil, err
+	}
+
+	return nodeList.Items, nil
+}
+
+func getNodeNames(nodes []corev1.Node) []string {
+	var nodeNames []string
+	for _, node := range nodes {
+		nodeNames = append(nodeNames, node.Name)
+	}
+
+	return nodeNames
+}
+
+func getMCPsFromMCOKubeletConfig(ctx context.Context, cli client.Client, mcoKubeletConfig mcov1.KubeletConfig) ([]*mcov1.MachineConfigPool, error) {
+	mcps := &mcov1.MachineConfigPoolList{}
+	var result []*mcov1.MachineConfigPool
+	if err := cli.List(ctx, mcps); err != nil {
+		return nil, err
+	}
+	for index := range mcps.Items {
+		mcp := &mcps.Items[index]
+
+		sel, err := metav1.LabelSelectorAsSelector(mcoKubeletConfig.Spec.MachineConfigPoolSelector)
+		if err != nil {
+			return nil, err
+		}
+		mcpLabels := labels.Set(mcp.Labels)
+		if sel.Matches(mcpLabels) {
+			result = append(result, mcp)
+		}
+
+	}
+	return result, nil
+}
+
+func validateKubeletConfig(nodeName string, kubeletConfiguration v1beta1.KubeletConfiguration) []validator.ValidationResult {
```

Comment:
I don't have strong feelings against reimplementing this logic, but why the `validate` pkg in `deployer` is not good enough?

Reply:
I was thinking about using a simplified version of `validator` as it seems to check some other things in deployer, and I also wanna make it check against a configurable "specification" ( or desired state ) instead of a hard coded one, but in the end I have used `deployer` `validator` package.

Reply:
checking against a configurable desired state is interesting, but I'd really start with a fixed set of hardcoded, always-needed settings. If `validator` does too much it's fine, just document it and reimplement the core checks (and/or send changes to `deployer`)